packages/xo-web/src/xo-app/vm/tab-advanced.js

```diff
@@ -288,6 +319,133 @@ class CoresPerSocket extends Component {
   }
 }
 
+class AddAclsModal extends Component {
+  get value() {
+    return this.state
+  }
+
+  _getPredicate = createSelector(
+    () => this.props.acls,
+    () => this.props.vm,
+    (acls, object) => ({ id: subject }) => !some(acls, { object, subject })
+  )
+
+  render() {
+    const { action, subjects } = this.state
+    return (
+      <form>
+        <div className='form-group'>
+          <SelectSubject
+            multi
+            onChange={this.linkState('subjects')}
+            predicate={this._getPredicate()}
+            value={subjects}
+          />
+        </div>
+        <div className='form-group'>
+          <SelectRole onChange={this.linkState('action')} value={action} />
+        </div>
+      </form>
+    )
+  }
+}
+
+const Acls = decorate([
+  addSubscriptions({
+    acls: subscribeAcls,
+    groups: subscribeGroups,
+    users: subscribeUsers,
+  }),
+  provideState({
+    effects: {
+      addAcls: () => (state, { acls, vm }) =>
+        confirm({
+          title: _('vmAddAcls'),
+          body: <AddAclsModal acls={acls} vm={vm} />,
+        })
+          .then(({ action, subjects }) => {
+            if (action == null || isEmpty(subjects)) {
+              return error(_('addAclsErrorTitle'), _('addAclsErrorMessage'))
+            }
+
+            return (
+              Promise.all(
+                map(subjects, subject =>
+                  addAcl({ subject, object: vm, action })
+                )
+              ),
+              noop
+            )
+          })
+          .catch(
+            err =>
+              err && error(_('addAclsErrorTitle'), err.message || String(err))
+          ),
+      removeAcl: (_, { currentTarget: { dataset } }) => (_, { vm: object }) =>
+        removeAcl({
+          action: dataset.action,
+          object,
+          subject: dataset.subject,
+        }),
+    },
+    computed: {
+      groups: (_, { groups }) => keyBy(groups, 'id'),
+      rawAcls: (_, { acls, vm }) => filter(acls, { object: vm }),
+      resolvedAcls: ({ rawAcls, users, groups }) =>
```

Comment:
Be aware that destructuring the state "uses" (ie calls the getter of) each property. So here, even if there aren't any ACLs, you're telling reaclette that you "need" `users` and `groups` so it will recompute them for you even though you sometimes don't actually need them.